runtime/src/accounts_db.rs

```diff
@@ -2807,58 +2839,125 @@ impl AccountsDB {
 
     // Force flush the cached roots, flush any unrooted frozen slots as well if there are
     // > MAX_CACHE_SLOTS of them.
-    pub fn force_flush_accounts_cache(&self) {
-        self.flush_accounts_cache(true);
+    pub fn force_flush_accounts_cache(&self, max_clean_root: Option<Slot>) {
+        self.flush_accounts_cache(true, max_clean_root);
     }
 
-    pub fn flush_accounts_cache_if_needed(&self) {
-        self.flush_accounts_cache(false);
+    pub fn flush_accounts_cache_if_needed(&self, max_clean_root: Option<Slot>) {
+        self.flush_accounts_cache(false, max_clean_root);
     }
 
-    fn flush_accounts_cache(&self, force_flush: bool) {
-        if !force_flush && self.accounts_cache.num_slots() <= MAX_CACHE_SLOTS {
-            return;
-        }
+    fn flush_rooted_slots_cache(
+        &self,
+        mut max_flush_root: Option<Slot>,
+        mut should_clean: Option<(&mut usize, &mut usize)>,
+    ) -> (usize, usize) {
+        if should_clean.is_some() {
+            max_flush_root = self.max_clean_root(max_flush_root);
+        }
+
+        // If there is a long running scan going on, this could prevent any cleaning
+        // past `max_flush_root`.
+        let cached_roots: BTreeSet<Slot> = self.accounts_cache.clear_roots(max_flush_root);
+        let mut written_accounts = HashSet::new();
+        // Iterate from highest to lowest so that we don't need to flush earlier
+        // outdated updates in earlier roots
+        let mut num_roots_flushed = 0;
+        let mut should_flush_f = |account_key: &Pubkey, account: &Account| {
+            // If a later root already wrote this account, no point
+            // in flushing it
+            if let Some((account_bytes_saved, num_accounts_saved)) = should_clean.as_mut() {
+                let should_flush = !written_accounts.contains(account_key);
+                if should_flush {
+                    written_accounts.insert(*account_key);
+                } else {
+                    **account_bytes_saved += account.data.len();
+                    **num_accounts_saved += 1;
```

Comment:
hehe, you want to see the actual savings? :)

Reply:
yeah was a bit curious :)